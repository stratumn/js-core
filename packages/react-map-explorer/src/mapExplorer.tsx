/*
  Copyright 2018 Stratumn SAS. All rights reserved.

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/

import { Segment } from '@stratumn/js-chainscript';
import React, { Component } from 'react';
import { hashToString } from './hash';
import { displayMap } from './mapExplorerDisplay';
import { IMapLoader } from './mapLoader';

/**
 * Props expected by the MapExplorer component.
 */
// tslint:disable-next-line:interface-name
export interface Props {
  /**
   * Map process.
   */
  process: string;

  /**
   * Id of the map you want to display.
   */
  mapId: string;

  /**
   * Hash of a known link that should be in the map.
   * If no link with this hash is currently loaded, the component will re-fetch
   * the map from the server.
   * You can use this prop to automatically refresh the map when creating new
   * links.
   */
  includeHash?: string;

  /**
   * Object that actually loads the whole map.
   */
  mapLoader: IMapLoader;

  /**
   * Callback when a segment is selected.
   */
  onSegmentSelected?: (s: Segment) => void;
}

/**
 * The MapExplorer component manages its own state.
 */
// tslint:disable-next-line:interface-name
export interface State {
  error?: Error;
  isLoaded: boolean;
  segments?: Segment[];
}

/**
 * A component to load and display Chainscript process maps.
 */
export class MapExplorer extends Component<Props, State> {
  public state: State = { isLoaded: false };
  private node?: SVGSVGElement;

  public async componentDidMount() {
    await this.loadMap();
    if (this.node && this.state.segments) {
      displayMap(this.node, this.state.segments, this.props.onSegmentSelected);
    }
  }

  public async componentDidUpdate(prevProps: Props) {
    if (this.shouldReloadMap(prevProps)) {
      await this.loadMap();
      if (this.node && this.state.segments) {
        displayMap(
          this.node,
          this.state.segments,
          this.props.onSegmentSelected
        );
      }
    }
  }

  public render() {
    if (!this.state.isLoaded) {
      return (
        <div>
          <p>Loading...</p>
        </div>
      );
    }

    if (this.state.error) {
      return (
        <div>
          <p>{this.state.error.message}</p>
        </div>
      );
    }

    return (
      <div>
        <svg
          ref={node => {
            if (node) {
              this.node = node;
            }
          }}
          width='800px'
          height='600px'
        />
      </div>
    );
  }

  private shouldReloadMap = (prevProps: Props): boolean => {
    // If the map we want to display has changed, refresh.
    if (
      prevProps.process !== this.props.process ||
      prevProps.mapId !== this.props.mapId
    ) {
      return true;
    }

    // If an unknown link should be included, refresh.
    if (
      this.props.includeHash &&
      prevProps.includeHash !== this.props.includeHash
    ) {
      if (!this.state.segments) {
        return true;
      }

      if (
        !this.state.segments.some(
          (s: Segment) => hashToString(s.linkHash()) === this.props.includeHash
        )
      ) {
        return true;
      }
    }

    return false;
  }

  private loadMap = async () => {
    try {
      const segments = await this.props.mapLoader.load(
        this.props.process,
        this.props.mapId
      );
      this.setState({
        isLoaded: true,
        segments
      });
    } catch (err) {
      this.setState({
        error: err,
        isLoaded: true
      });
    }
  }
}
