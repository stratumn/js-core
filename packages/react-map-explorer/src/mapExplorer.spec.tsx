/*
  Copyright 2018 Stratumn SAS. All rights reserved.

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/

import { shallow } from 'enzyme';
import React from 'react';
import { MapWithoutRefs, TestMapId, TestProcess } from '../test/fixtures/maps';
import { MapExplorer, State } from './mapExplorer';

describe('map explorer', () => {
  const mapLoader = {
    load: jest.fn()
  };
  const testProps = {
    mapId: TestMapId,
    mapLoader,
    process: TestProcess
  };

  beforeEach(() => {
    mapLoader.load.mockClear();
  });

  it('renders a loading spinner', () => {
    const wrapper = shallow(<MapExplorer {...testProps} />);

    expect(
      wrapper
        .find('p')
        .first()
        .text()
    ).toEqual('Loading...');
  });

  it('renders an error', async () => {
    mapLoader.load.mockRejectedValueOnce(new Error('Network Failure'));
    const wrapper = shallow(<MapExplorer {...testProps} />);

    await new Promise(resolve => setImmediate(resolve));
    wrapper.update();

    expect(mapLoader.load).toHaveBeenCalled();
    expect(
      wrapper
        .find('p')
        .first()
        .text()
    ).toEqual('Network Failure');
  });

  it('loads the map segments', () => {
    shallow(<MapExplorer {...testProps} />);

    expect(mapLoader.load).toHaveBeenCalled();
    expect(mapLoader.load).toHaveBeenCalledWith(
      testProps.process,
      testProps.mapId
    );
  });

  it('renders map segments count', async () => {
    mapLoader.load.mockResolvedValueOnce(MapWithoutRefs);
    const wrapper = shallow(<MapExplorer {...testProps} />);

    await new Promise(resolve => setImmediate(resolve));
    wrapper.update();

    expect(
      wrapper
        .find('h2')
        .first()
        .text()
    ).toEqual('2 segments found');
  });

  it('renders a list of segments', async () => {
    const segmentSelected = jest.fn();
    mapLoader.load.mockResolvedValueOnce(MapWithoutRefs);
    const wrapper = shallow(
      <MapExplorer {...testProps} onSegmentSelected={segmentSelected} />
    );

    await new Promise(resolve => setImmediate(resolve));
    wrapper.update();

    expect(wrapper.find('ul').children()).toHaveLength(2);

    wrapper
      .find('li')
      .first()
      .simulate('click');
    expect(segmentSelected).toHaveBeenCalled();
    expect(segmentSelected).toHaveBeenCalledWith(MapWithoutRefs[0]);
  });

  it('reloads when mapId changes', () => {
    mapLoader.load.mockResolvedValue(MapWithoutRefs);
    const wrapper = shallow(<MapExplorer {...testProps} />);
    expect(mapLoader.load).toHaveBeenCalled();

    wrapper.setProps({ mapId: 'other_map', process: 'other_process' });
    expect(mapLoader.load).toHaveBeenCalledWith('other_process', 'other_map');
  });

  it('reloads when segments where not loaded', async () => {
    mapLoader.load.mockResolvedValue([]);
    const wrapper = shallow(<MapExplorer {...testProps} />);
    expect(mapLoader.load).toHaveBeenCalledTimes(1);

    wrapper.setProps({ includeHash: '1234' });
    expect(mapLoader.load).toHaveBeenCalledTimes(2);
  });

  it('reloads when new segment should be included', async () => {
    mapLoader.load.mockResolvedValue(MapWithoutRefs);
    const wrapper = shallow(<MapExplorer {...testProps} />);
    expect(mapLoader.load).toHaveBeenCalledTimes(1);

    await new Promise(resolve => setImmediate(resolve));
    wrapper.update();

    wrapper.setProps({ includeHash: '1234' });
    expect(mapLoader.load).toHaveBeenCalledTimes(2);
  });
});
