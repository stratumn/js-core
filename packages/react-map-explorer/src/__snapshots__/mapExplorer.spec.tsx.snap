// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`map explorer renders svg elements for each segment: map-without-refs 1`] = `"<div><svg width=\\"800px\\" height=\\"600px\\"><g transform=\\"translate(50,250)\\"><g class=\\"nodes\\"><rect class=\\"node\\" x=\\"0\\" y=\\"0\\" width=\\"100\\" height=\\"100\\" style=\\"fill: steelblue;\\"></rect><rect class=\\"node\\" x=\\"150\\" y=\\"0\\" width=\\"100\\" height=\\"100\\" style=\\"fill: steelblue;\\"></rect><text x=\\"10\\" y=\\"50\\">1e1c49...</text><text x=\\"160\\" y=\\"50\\">8544bc...</text></g><g class=\\"links\\"><line class=\\"link\\" x1=\\"100\\" y1=\\"50\\" x2=\\"150\\" y2=\\"50\\" style=\\"stroke: #ccc; stroke-width: 1px;\\"></line></g></g></svg></div>"`;
